File: backend/app/api/v1/documents.py
from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session

from app.api.deps import get_current_user
from app.db.session import get_db
from app.models.document import DocumentVerification
from app.models.user import User
from app.schemas.application import DocumentVerifyIn, VerificationResult

router = APIRouter(prefix="/documents", tags=["documents"])


@router.post("/verify", response_model=VerificationResult)
def verify_document(
    payload: DocumentVerifyIn,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> VerificationResult:
    """
    Receives only boolean checks + a masked identifier from the browser —
    never a raw document image or full ID number. OCR happens client-side
    via Tesseract.js; this endpoint just records the verification outcome.
    """
    status = "verified" if all(payload.checks.values()) else "failed"

    record = DocumentVerification(
        user_id=current_user.id,
        document_type=payload.doc_type,
        verification_status=status,
        masked_identifier=payload.masked_id,
    )
    db.add(record)
    db.commit()
    db.refresh(record)

    return VerificationResult(
        id=record.id,
        document_type=record.document_type,
        verification_status=record.verification_status,
        masked_identifier=record.masked_identifier,
        checked_at=record.checked_at,
    )

File: frontend/lib/documents.ts
/**
 * Client binding for MY (Member 3) POST /documents/verify endpoint.
 * See backend/app/api/v1/documents.py and backend/app/schemas/application.py
 * (DocumentVerifyIn / VerificationResult).
 *
 * Builds its payload from the document-readiness checker's existing masked
 * result (see lib/document-readiness/storage.ts's toStoredResult, same
 * "never send raw OCR text or full numbers" rule) — does not touch any
 * OCR/validation logic itself.
 */

import { apiPost } from "@/lib/api-client"
import type { DocumentReadinessResult } from "@/lib/document-readiness/types"

export interface DocumentVerifyIn {
  doc_type: string
  checks: Record<string, boolean>
  masked_id?: string | null
}

export interface VerificationResult {
  id: string
  document_type: string
  verification_status: string
  masked_identifier: string | null
  checked_at: string
}

export function toVerifyPayload(result: DocumentReadinessResult): DocumentVerifyIn {
  const hasCriticalIssue = result.issues.some((issue) => issue.severity === "critical")
  const hasNameMismatch = result.issues.some((issue) => issue.code === "name_mismatch")
  const lastFour = result.extractedFields.aadhaarLastFour ?? result.extractedFields.accountNumberLastFour

  return {
    doc_type: result.documentType,
    checks: {
      ocr_readable: result.status !== "unclear" && result.status !== "error",
      no_critical_issues: !hasCriticalIssue,
      name_matches: !hasNameMismatch,
    },
    masked_id: lastFour ? `••••${lastFour}` : null,
  }
}

export function verifyDocument(result: DocumentReadinessResult): Promise<VerificationResult> {
  return apiPost<VerificationResult>("/documents/verify", toVerifyPayload(result))
}
